File: behavioral/typescript/chain_of_responsibility_structured/approvers.ts
import { Approver } from './interfaces';

export class TeamLead extends Approver {
  approveRequest(amount: number): void {
    if (amount <= 1000) {
      console.log(`TeamLead: Approved request of $${amount}`);
    } else {
      console.log('TeamLead: Amount exceeds limit, passing to manager...');
      this.passRequest(amount);
    }
  }
}

export class ManagerApprover extends Approver {
  approveRequest(amount: number): void {
    if (amount <= 5000) {
      console.log(`Manager: Approved request of $${amount}`);
    } else {
      console.log('Manager: Amount exceeds limit, passing to director...');
      this.passRequest(amount);
    }
  }
}

export class Director extends Approver {
  approveRequest(amount: number): void {
    if (amount <= 20000) {
      console.log(`Director: Approved request of $${amount}`);
    } else {
      console.log('Director: Amount exceeds limit, passing to CEO...');
      this.passRequest(amount);
    }
  }
}

export class CEO extends Approver {
  approveRequest(amount: number): void {
    console.log(`CEO: Approved request of $${amount} (final approver)`);
  }
}

File: behavioral/dart/chain_of_responsibility_structured/approvers.dart
import 'interfaces.dart';

class TeamLead extends Approver {
  @override
  void approveRequest(double amount) {
    if (amount <= 1000) {
      print('TeamLead: Approved request of \$${amount}');
    } else {
      print('TeamLead: Amount exceeds limit, passing to manager...');
      passRequest(amount);
    }
  }
}

class ManagerApprover extends Approver {
  @override
  void approveRequest(double amount) {
    if (amount <= 5000) {
      print('Manager: Approved request of \$${amount}');
    } else {
      print('Manager: Amount exceeds limit, passing to director...');
      passRequest(amount);
    }
  }
}

class Director extends Approver {
  @override
  void approveRequest(double amount) {
    if (amount <= 20000) {
      print('Director: Approved request of \$${amount}');
    } else {
      print('Director: Amount exceeds limit, passing to CEO...');
      passRequest(amount);
    }
  }
}

class CEO extends Approver {
  @override
  void approveRequest(double amount) {
    print('CEO: Approved request of \$${amount} (final approver)');
  }
}
